### PR TITLE
Remove dead code from CSharpCodeAnalysis project

### PR DESCRIPTION
**Pull request by alrz**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Nice.  Did you use a particular tool to find this?  Would love to see the same applied to the Workspace and above layers!

**Comment by alrz** [comment]
I'm using R# but it's fairly tedious cuz it doesn't do fix-all. Would love to see this as a Roslyn analyzer since it comes with fix-all and all. :)

**Comment by CyrusNajmabadi** [comment]
Sounds like a really good analyzer for Roslyn-Analyzers.

**Comment by Pilchie** [comment]
Yes.  This is high on @kuhlenh's list.  The tough part is providing a decent set of "roots" for the code in cases where it isn't just a program with a main method.

**Comment by alrz** [comment]
> The tough part is providing a decent set of "roots" for the code in cases where it isn't just a program with a main method.

I think we should give a compiler-produced warning just like for unused variables, fields, etc. that's fairly easier for the compiler to figure this out. From there, writing an analyzer is rather a trivial task. (related: warning waves)

**Comment by Pilchie** [comment]
We do give a warning for unused locals, and I agree that it could be expanded to private methods, and maybe internal ones of assemblies that don't have an InternalsVisibleTo.  Unfortunately, the problem is that without some heuristics about entry points, this sort of feature can never flag:

* Public methods
* Internal methods of assemblies with an InternalsVisibleTo attribute
* Methods that implement an interface
* Overrides of virtual methods
* etc.

In other words, there is some value to the feature based on a strict analysis, but not as much as there could be.

**Comment by alrz** [comment]
I wouldn't set the expectation that high, a more realistic approach would be to flag members that are "definitely" out of use. I'm just saying that the compiler can get there faster while it's walking symbols to produce other diagnostics like unused fields, etc.

**Comment by kuhlenh** [comment]
@Pilchie the roslyn-analyzer repo also has this for parameters (notifies you if you have a parameter that you never use).

**Comment by gafter** [comment]
Please retarget to 15.6.

**Comment by tmat** [comment]
@ManishJayaswal Perhaps the roots could be test projects. It might also be useful to combine with test coverage analysis. If a method has no line covered then that's a red flag (it may still be reachable) and on the other hand if a method is not statically reachable but has line covered then it can't be removed (it was invoked e.g. thru reflection).

**Comment by tmat** [comment]
Also note that the static analysis is extensively done in ILLinker and .NET Native.

**Comment by ManishJayaswal** [comment]
@shyamnamboodiripad can you take a look at what @tmat is recommending above?

**Comment by alrz** [comment]
@tmat Why this is different from fields? The compiler already give warnings for unused fields at certain circumstances. Why can't we apply the same to other members?

**Comment by jcouv** [comment]
@dotnet-bot retest windows_release_unit32_prtest please

**Comment by jcouv** [comment]
retest windows_release_unit64_prtest please

**Comment by jcouv** [comment]
retest windows_debug_unit32_prtest please

**Comment by jcouv** [comment]
retest windows_build_correctness_prtest please

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. Community contribution to remove dead code, for 15.6.
